src/extension/provider/statefulAuth.ts

```diff
@@ -0,0 +1,579 @@
+import * as crypto from 'node:crypto'
+
+import {
+  authentication,
+  AuthenticationProvider,
+  Disposable,
+  env,
+  EventEmitter,
+  ExtensionContext,
+  ProgressLocation,
+  Uri,
+  window,
+  AuthenticationSession,
+  AuthenticationProviderAuthenticationSessionsChangeEvent,
+  Event,
+} from 'vscode'
+import { v4 as uuid } from 'uuid'
+import fetch from 'node-fetch'
+
+import { getIdpConfig, getRunmeAppUrl } from '../../utils/configuration'
+import { AuthenticationProviders } from '../../constants'
+import { RunmeUriHandler } from '../handler/uri'
+
+const AUTH_NAME = 'Stateful'
+const SESSIONS_SECRET_KEY = `${AuthenticationProviders.Stateful}.sessions`
+
+interface TokenInformation {
+  access_token: string
+  refresh_token: string
+  expires_in: number
+}
+
+interface StatefulAuthSession extends AuthenticationSession {
+  refreshToken: string
+  expiresIn: number
+}
+
+// Interface declaration for a PromiseAdapter
+interface PromiseAdapter<T, U> {
+  // Function signature of the PromiseAdapter
+  (
+    // Input value of type T that the adapter function will process
+    value: T,
+    // Function to resolve the promise with a value of type U or a promise that resolves to type U
+    resolve: (value: U | PromiseLike<U>) => void,
+    // Function to reject the promise with a reason of any type
+    reject: (reason: any) => void,
+  ): any // The function can return a value of any type
+}
+
+const passthrough = (value: any, resolve: (value?: any) => void) => resolve(value)
+
+export class StatefulAuthProvider implements AuthenticationProvider, Disposable {
+  #disposables: Disposable[] = []
+  #pendingStates: string[] = []
+  #codeVerfifiers = new Map<string, string>()
+  #scopes = new Map<string, string[]>()
+  #uriHandler: RunmeUriHandler
+  #codeExchangePromises = new Map<
+    string,
+    { promise: Promise<TokenInformation>; cancel: EventEmitter<void> }
+  >()
+
+  readonly #onSessionChange = this.register(
+    new EventEmitter<AuthenticationProviderAuthenticationSessionsChangeEvent>(),
+  )
+
+  constructor(
+    private readonly context: ExtensionContext,
+    uriHandler: any,
```

Comment:
why `any`?

Reply:
I changed before, Relived it again :sweat_smile: